File: behaviorial/interpreter/python/nonterminal_expression.py
from contextvars import Context
from abstract_expression import Expression


class Add(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self, context: Context):
        return self.left.interpret(context) + self.right.interpret(context)


class Multiply(Expression):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    def interpret(self, context: Context):
        return self.left.interpret(context) * self.right.interpret(context)

File: behaviorial/interpreter/typescript/nonterminal-expression.ts
import { Context } from './context';
import { Expression } from './abstract-expression';

export class Add extends Expression {
	private left: Expression;
	private right: Expression;

	constructor(left: Expression, right: Expression) {
		super();
		this.left = left;
		this.right = right;
	}

	interpret(context: Context): number {
		return this.left.interpret(context) + this.right.interpret(context);
	}
}

export class Multiply extends Expression {
	private left: Expression;
	private right: Expression;

	constructor(left: Expression, right: Expression) {
		super();
		this.left = left;
		this.right = right;
	}

	interpret(context: Context): number {
		return this.left.interpret(context) * this.right.interpret(context);
	}
}
